aiokafka/producer.py

```diff
@@ -255,25 +303,130 @@ def send(self, topic, value=None, key=None, partition=None):
 
         tp = TopicPartition(topic, partition)
-        leader = self._metadata.leader_for_partition(tp)
-        if leader is None:
-            raise NotLeaderForPartitionError()
-        elif leader == -1:
-            raise LeaderNotAvailableError()
-
         log.debug("Sending (key=%s value=%s) to %s", key, value, tp)
-        msg = Message(value_bytes, key=key_bytes)
-        buf = MessageSetBuffer(
-            io.BytesIO(), self._batch_size, self._compression_type)
-        buf.append(0, msg)
-        buf.close()
-        request = ProduceRequest(
-            required_acks=self._acks,
-            timeout=self._request_timeout_ms,
-            topics=[(topic, [(partition, buf.buffer())])])
-
-        response = yield from self.client.send(leader, request)
-        return response
+
+        fut = yield from self._message_accumulator.add_message(
+            tp, key_bytes, value_bytes)
+        return fut
+
+    @asyncio.coroutine
+    def _sender_routine(self):
+        """backgroud task that sends message batches to Kafka brokers"""
+        tasks = set()
+        try:
+            while True:
+                batches, unknown_leaders_exist = \
+                    self._message_accumulator.drain_by_nodes(
+                        ignore_nodes=self._in_flight)
+
+                # create produce task for every batch
+                for node_id, batches in batches.items():
+                    task = ensure_future(
+                        self._send_produce_req(node_id, batches),
+                        loop=self._loop)
+                    tasks.add(task)
+
+                if unknown_leaders_exist:
+                    # we have at least one unknown partition's leader,
+                    # try to update cluster metadata
+                    yield from self.client.force_metadata_update()
+
+                if tasks:
+                    # wait when at least one of produce task is finished
+                    _, tasks = yield from asyncio.wait(
+                        tasks, return_when=asyncio.FIRST_COMPLETED,
+                        loop=self._loop)
+                elif not unknown_leaders_exist:
+                    # we have no tasks and all partition's leader are known
+                    # so just wait new data (at least one message)
+                    yield from self._message_accumulator.wait_data()
+                else:
+                    # we have no tasks and some leaders for partitions are
+                    # unknown. In this case we should sleep some time before
+                    # next drain/metadata_update
+                    yield from asyncio.sleep(
+                        self._send_backoff_time, loop=self._loop)
+        except asyncio.CancelledError:
+            pass
+        except Exception:  # noqa
+            log.error("Unexpected error in sender routine", exc_info=True)
+        finally:
+            # wait in-progress produce tasks before close
+            if tasks:
```

Comment:
This may result is strange behaviour. We don't fail all requests. We just wait for them...